tools/scoring/score/dimensions/entity_type_identification.py

```diff
@@ -0,0 +1,169 @@
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the License);
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an AS IS BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+""" Core component """
+
+from score.dimensions.dimension import Dimension
+from score.scorer_types import DeserializedFile, EntityInstance, CloudDeviceId, PointsVirtualList
+from score.constants import FileTypes
+
+from typing import Set
+
+PROPOSED, SOLUTION = FileTypes
+
+
+class EntityTypeIdentification(Dimension):
+  """
+  Quantifies whether the proposed file
+  assigned the correct DBO type to each entity.
+  """
+  def _isolate_entities_virtual(self,
+                                file: DeserializedFile) -> Set[EntityInstance]:
+    return set(
+        filter(self.is_entity_canonical,
+               filter(self.is_entity_virtual, file.values())))
+
+  def _fetch_points_virtual(
+      self, file: DeserializedFile,
+      entities_virtual: Set[EntityInstance]) -> PointsVirtualList:
+    # For each canonically typed virtual entity
+    # create a tuple containing a set and the entity's type.
+    # (The type is required by the matching algo for a different dimension.)
+    # For each link in the entity,
+    # if the field exists at the source
+    # add its raw field name to the set.
+    # Filter out sets which have no items
+    # and sort by number of fields represented in descending order.
+    # TODO: move the filtering/sorting functionality to add clarity.
+    return sorted(
+        list(
+            filter(
+                lambda tup: len(tup[0]) > 0,  # (raw field names, entity type)
+                [(set(file[link.source].translation[target_field].raw_field_name
```

Comment:
can you please comment inline here, I dont follow what this is doing, it is hard to follow.

Reply:
It's commented above.

Reply:
sure but I dont follow it maybe use a helper or apply the TODO?